blockchain/common_test.go

```diff
@@ -56,55 +56,39 @@ func isSupportedDbType(dbType string) bool {
 	return false
 }
 
-// createTestDatabase creates a test database with the provided database name
-// and database type for the given network.
-func createTestDatabase(dbName string, dbType string, net wire.CurrencyNet) (database.DB, func(), error) {
+// createTestDatabase creates a test database with the provided database type
+// for the given network.
+func createTestDatabase(t testing.TB, dbType string, net wire.CurrencyNet) (database.DB, error) {
 	// Handle memory database specially since it doesn't need the disk specific
 	// handling.
 	var db database.DB
-	var teardown func()
 	if dbType == "memdb" {
 		ndb, err := database.Create(dbType)
 		if err != nil {
-			return nil, nil, fmt.Errorf("error creating db: %w", err)
+			return nil, fmt.Errorf("error creating db: %w", err)
 		}
 		db = ndb
-
-		// Setup a teardown function for cleaning up.  This function is returned to
-		// the caller to be invoked when it is done testing.
-		teardown = func() {
-			db.Close()
-		}
 	} else {
 		// Create the directory for the test database.
-		dbPath, err := os.MkdirTemp("", dbName)
-		if err != nil {
-			err := fmt.Errorf("unable to create test db path: %w",
-				err)
-			return nil, nil, err
-		}
+		dbPath := t.TempDir()
 
 		// Create the test database.
 		ndb, err := database.Create(dbType, dbPath, net)
 		if err != nil {
-			os.RemoveAll(dbPath)
-			return nil, nil, fmt.Errorf("error creating db: %w", err)
+			return nil, fmt.Errorf("error creating db: %w", err)
 		}
 		db = ndb
-
-		// Setup a teardown function for cleaning up.  This function is returned to
-		// the caller to be invoked when it is done testing.
-		teardown = func() {
-			db.Close()
-			os.RemoveAll(dbPath)
-		}
 	}
+	t.Cleanup(func() {
+		db.Close()
+	})
 
```

Comment:
@davecgh Reply for https://github.com/decred/dcrd/pull/2902/files/22353b7131cb052a72cc87742fc11532429c6aea#r828700005

The teardown `db.Close()` is moved into `t.Cleanup`. This ensures that `db.Close()` is always called when the test finishes.

Reply:
Indeed.  I was calling it out because I thought the `t.Cleanup` required Go 1.18, which we can't require yet and without that, it would lead to not closing the db.  However, as you correctly pointed out, `t.Cleanup` only requires Go 1.14 and so the comment regarding not being closed is no longer relevant with the `t.Cleanup` code staying.

Reply:
Although I am seeing a failure related to this on Windows:

```
--- FAIL: TestProcessLogic (0.70s)
    process_test.go:533: Failed to setup chain instance: The process cannot access the file because it is being used by another process.
```

Reply:
The culprit looks like the fact that the database for the shared generator (created once by `genSharedProcessTestBlocks`) must be closed and removed prior to returning from that function, thereby only leaving the in-memory block structure which is then shared among various tests which create their own chain instance and backing database.

There are probably a few way to address it, but I suspect what would probably be the fewest changes would be to rename the exist `chainSetup` code that returns a teardown func to something else and then create a new `chainSetup` wrapper that calls the teardown func via `t.Cleanup`.  That way, all of the rest of the code updates will remain as they are and make use of the automatic cleanup, but the shared generator paths can instead make use of the `chainSetup` that returns the teardown func and call it as needed.

Reply:
This commit 9c56f8d7e8b574b47a30580359dd2a423d5ef93e should fix the issue. `TestProcessLogic` was failing because `createTestUtxoDatabase` was called by `genSharedProcessTestBlocks` and `newChaingenHarnessWithGen`. If we give `createTestUtxoDatabase` a static database name then there will be conflict. 

I don't have a Windows machine so I am running the tests on the GitHub-hosted Windows runner on my fork.

Reply:
That works too.  It does leave the temporary database used to generated the shared blocks around for longer than is strictly necessary, but given it's for testing purposes, that's not a big deal.

Confirmed the updates work properly on Windows now too.